include/unifex/unstoppable_baton.hpp

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/config.hpp>
+#include <unifex/receiver_concepts.hpp>
+
+#include <atomic>
+#include <type_traits>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+
+namespace _ub {
+
+struct _op_base;
+
+template <typename Receiver>
+struct _operation {
+  struct type;
+};
+
+template <typename Receiver>
+using operation = typename _operation<Receiver>::type;
```

Comment:
Some of the code below would be simpler if this template alias was defined as
```
template <typename Receiver>
using operation = typename _operation<std::decay_t<Receiver>>::type;
```

Is there a preference one way or the other?

Reply:
I like it one way. Lewis likes it the other. :-P

Reply:
Sounds like I should follow the established convention, then.  I think the convention is not to embed `decay_t` in this alias, right?

Reply:
I think mostly the aliases do the de-qualification, since I did most of the ADL isolation work.

Reply:
I was trying to find a pattern to follow and ran into the construction of the operation state in [`let_with`](https://github.com/facebookexperimental/libunifex/blob/master/include/unifex/let_with.hpp#L82).  Because that operation state can be instantiated with some `Receiver&` type, [this constructor](https://github.com/facebookexperimental/libunifex/blob/master/include/unifex/let_with.hpp#L96) takes either an lvalue or rvalue reference.  In contrast, my operation state takes its receiver by value and `move`s it into place.

It feels like there should be a policy I follow here and it's not obvious to me what the policy should be.  Is there already a consensus?  The approach in `let_with` potentially leads to more distinct template instantiations because it could be instantiated with both `Receiver` and `Receiver&`, but it probably results in a more efficient constructor.

Reply:
Personally, I think taking by value and moving it into `connect` is perfectly fine.

Reply:
> I think mostly the aliases do the de-qualification, since I did most of the ADL isolation work.

I tended to prefer to leave the decay out of the alias so that I can tell its doing the right thing at point of use.
If I the decay is in the alias then I always had to scroll up to see the definition of the alias to check which template args implicitly had what transformation performed to them.

Having the aliases be straight aliases I feel separates the concerns better. The aliases are there to simplify use of the ADL-isolation wrappers and that only.

Reply:
> Personally, I think taking by value and moving it into connect is perfectly fine.

Yes, I think that's probably fine. The vast majority of receivers end up just being a pointer or two anyway which should be pretty cheap to copy and/or pretty inlinable.